asab/library/providers/zookeeper.py

```diff
@@ -30,9 +31,21 @@ def __init__(self, app, path):
 			config_section_name='',
 			z_path=self.Path
 		)
-		self.Zookeeper = self.ZookeeperContainer.ZooKeeper
+		self.Zookeeper = None
+		self.App.PubSub.subscribe("ZooKeeperContainer.started!", self._on_zk_ready)
+
+	async def finalize(self, app):
+		# close client
+		await self.Zookeeper._stop()
+
+	async def _on_zk_ready(self, event_name, zkcontainer):
+		if zkcontainer == self.ZookeeperContainer:
+			self.Zookeeper = self.ZookeeperContainer.ZooKeeper
 
 	async def read(self, path):
+		if self.Zookeeper is None:
+			return
```

Comment:
You should warn (into logging) when this happens ... otherwise you just set a trap to a developer.